src/freenet/keys/SSKBlock.java

```diff
@@ -137,22 +138,32 @@ public SSKBlock(byte[] data, byte[] headers, NodeSSK nodeKey, boolean dontVerify
 			System.arraycopy(headers, x, bufS, 0, SIG_S_LENGTH);
 			x+=SIG_S_LENGTH;
 
-	        MessageDigest md = SHA256.getMessageDigest();
-			md.update(data);
-			byte[] dataHash = md.digest();
-			// All headers up to and not including the signature
-			md.update(headers, 0, headersOffset + ENCRYPTED_HEADERS_LENGTH);
-			// Then the implicit data hash
-			md.update(dataHash);
-			// Makes the implicit overall hash
-			byte[] overallHash = md.digest();
-			SHA256.returnMessageDigest(md);
+			MessageDigest md = null;
+			byte[] overallHash;
+			try {
+				md = SHA256.getMessageDigest();
+				md.update(data);
+				byte[] dataHash = md.digest();
+				// All headers up to and not including the signature
+				md.update(headers, 0, headersOffset + ENCRYPTED_HEADERS_LENGTH);
+				// Then the implicit data hash
+				md.update(dataHash);
+				// Makes the implicit overall hash
+				overallHash = md.digest();
+			} finally {
+				SHA256.returnMessageDigest(md);
+			}
 			
 			// Now verify it
-			NativeBigInteger r = new NativeBigInteger(1, bufR);
-			NativeBigInteger s = new NativeBigInteger(1, bufS);
-			if(!(DSA.verify(pubKey, new DSASignature(r, s), new NativeBigInteger(1, overallHash), false) ||
-					(DSA.verify(pubKey, new DSASignature(r, s), new NativeBigInteger(1, overallHash), true)))) {
+			BigInteger r = new BigInteger(1, bufR);
+			BigInteger s = new BigInteger(1, bufS);
+			DSASigner dsa = new DSASigner();
+			dsa.init(false, new DSAPublicKeyParameters(pubKey.getY(), Global.getDSAgroupBigAParameters()));
+
+			// We probably don't need to try both here... but that's what the legacy code was doing.
```

Comment:
Q is invariant in our case (we always use the same group)... and from what I can tell we always truncate... whereas I think that we don't have to... do we want to keep the same behaviour? The madness has started in 567bf11a954edc31f3b6d4348782792fe7d5bae5 ; Do we care about the content that was inserted before? If we don't we might be able to safe half the cpu cycles here.

For the record, we don't need "m < q"!
It's the opposite: "N must be less than or equal to the output length of the hash H." according to https://en.wikipedia.org/wiki/Digital_Signature_Algorithm